### PR TITLE
feat: implement theme

### PR DESCRIPTION
**Pull request by loido**

### CODE REVIEW

**Review by anhle10051996 — CHANGES_REQUESTED** [review]

File: src/Hello.tsx [hunk 1]
```diff
@@ -2,63 +2,42 @@
...
-  button: {
-    flex: 1,
-    paddingVertical: 0,
-  },
-  greeting: {
-    color: '#999',
-    fontWeight: 'bold',
-  },
-})
+const Demo = styled(View)`
```
[the remark this answers is not in the document]
Comment (loido) [reply src/Hello.tsx]: check again, please

File: src/core/color-mode/type.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+export interface ColorModeOptions {
+  initialColorMode?: ColorMode
+  useSystemColorMode?: boolean
+}
+
+export interface IColorModeContextProps {
+  colorMode: ColorMode
+  toggleColorMode: () => void
+  setColorMode: (value: any) => void
```
[the remark this answers is not in the document]
Comment (loido) [reply src/core/color-mode/type.ts]: check again, please

File: src/core/BaseProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+export interface BaseProviderProps {
+  children?: React.ReactNode
+  theme?: ITheme
+}
+
+const BaseProvider = ({children, theme = defaultTheme}: BaseProviderProps) => {
+  const [colorModeValue, setColorModeValue] = useState<ColorMode>(theme?.config.initialColorMode)
+
+  const isLight = useMemo(() => {
+    return colorModeValue === 'light'
```
Comment (honghoangsts) [inline src/core/BaseProvider.tsx:20]: no magic string.

File: src/core/BaseProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  }, [colorModeValue])
+
+  const darkTheme = useMemo(() => {
+    return {...theme, colors: {...theme.colors, ...theme.darkColors}}
+  }, [theme])
+
+  const newTheme = useMemo(() => {
+    return isLight ? theme : darkTheme
```
Comment (honghoangsts) [inline src/core/BaseProvider.tsx:28]: We agree use only one themes .

File: src/theme/base/typography.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    hairline: 100,
+    thin: 200,
+    light: 300,
+    normal: 400,
+    medium: 500,
+    semibold: 600,
+    bold: 700,
+    extrabold: 800,
+    black: 900,
+    extraBlack: 950,
```
Comment (honghoangsts) [inline src/theme/base/typography.ts:12]: Does fontWeight work for android?

File: src/theme/base/typography.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    lg: 18,
+    xl: 20,
+    '2xl': 24,
+    '3xl': 30,
+    '4xl': 36,
+    '5xl': 48,
+    '6xl': 60,
+    '7xl': 72,
+    '8xl': 96,
+    '9xl': 128,
```
Comment (honghoangsts) [inline src/theme/base/typography.ts:33]: We should think again about naming for fontSize

**Review by honghoangsts — CHANGES_REQUESTED** [review]

File: package.json [hunk 7]
```diff
@@ -55,6 +55,7 @@
     "@types/react": "^18.0.26",
     "@types/react-native": "^0.70.0",
     "@types/react-test-renderer": "^18.0.0",
+    "@types/styled-components": "^5.1.26",
```
Comment (loc-nguyenthien) [inline package.json:58]: I think we also need `@types/styled-components-react-native`
https://styled-components.com/docs/api#typescript

File: src/theme/theme.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import type {ColorModeOptions} from 'src/core/color-mode/type'
```
Comment (loc-nguyenthien) [inline src/theme/theme.ts:1]: Please check this import is valid

File: src/hooks/useTheme.tsx [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import type {ITheme} from 'src/theme'
```
Comment (loc-nguyenthien) [inline src/hooks/useTheme.tsx:1]: Please check this import is valid

**Review by loc-nguyenthien — CHANGES_REQUESTED** [review]

File: package.json (hunk 7 above)
Comment (loc-nguyenthien) [reply package.json:58]: updated

File: src/hooks/useTheme.tsx (hunk 9 above)
Comment (loc-nguyenthien) [reply src/hooks/useTheme.tsx]: updated

File: src/theme/theme.ts (hunk 8 above)
Comment (loc-nguyenthien) [reply src/theme/theme.ts]: updated

File: src/theme/base/typography.ts (hunk 5 above)
Comment (loido) [reply src/theme/base/typography.ts:12]: it is work both android and ios

File: src/theme/theme.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import type {ColorModeOptions} from '../core/color-mode/type'
+import base from './base'
+
+const config: ColorModeOptions = {
+  useSystemColorMode: false,
+  initialColorMode: 'light',
```
Comment (TranQuangPhi) [inline src/theme/theme.ts:6]: I think the better way would be to define an enum for this

File: src/core/BaseProvider.tsx (hunk 4 above)
Comment (loido) [reply src/core/BaseProvider.tsx:28]: we can use only theme

File: src/theme/theme.ts (hunk 10 above)
Comment (loido) [reply src/theme/theme.ts:6]: but this initialColorMode is use ColorMode's type

File: src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,5 @@
+import {extendTheme} from './core/extendTheme'
+import BaseProvider from './core/BaseProvider'
+import Hello from './Hello'
+
+export {extendTheme, BaseProvider, Hello}
```
Comment (ThinhKimVo) [inline src/index.tsx:5]: `Hello` ?

File: src/theme/base/typography.ts (hunk 5 above)
Comment (vunguyendhSTS) [reply src/theme/base/typography.ts:12]: fontWeight probably doesn't work on Android. we just should use fontFamily.

File: src/index.tsx (hunk 11 above)
Comment (dovanloidev) [reply src/index.tsx:5]: Hello component is an example for theme

File: package.json [hunk 12]
```diff
@@ -2,8 +2,8 @@
   "name": "rn-base-component",
   "version": "0.0.2",
   "description": "test",
-  "main": "lib/commonjs/index",
-  "module": "lib/module/index",
+  "main": "src/index",
+  "module": "src/index",
```
Comment (ngochuyduong) [inline package.json:6]: Why change this?

**Review by ngochuyduong — CHANGES_REQUESTED** [review]

File: package.json (hunk 12 above)
Comment (loido) [reply package.json:6]: export components for test

File: src/index.tsx (hunk 11 above)
Comment (ThinhKimVo) [reply src/index.tsx:5]: **WelcomeScreen** should be more make sense

File: src/theme/base/typography.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    black: 900,
+    extraBlack: 950,
+  },
+  fonts: {
+    heading: undefined,
+    body: undefined,
+    mono: undefined,
+  },
+  fontSizes: {
+    '2xs': 10,
```
Comment (hoangSTS) [inline src/theme/base/typography.ts:20]: why don't you use xxs instead of '2xs'?

File: src/Hello.tsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  return <Demo />
+}
+
+export default Hello
+
+const Demo = styled.View`
+  width: 100px;
+  height: 100px;
+  background-color: ${(props: {theme: ITheme}) => props?.theme?.colors?.cardPrimaryBackground};
+`
```
Comment (honghoangsts) [inline src/Hello.tsx:43]: Don't use CSS

File: src/Hello.tsx [hunk 15]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Created by NL on 04/01/2023.
+ */
+// src/Hello.tsx
+import React from 'react'
+import styled from 'styled-components/native'
+import {extendTheme} from './core/extendTheme'
+import BaseProvider from './core/BaseProvider'
+import {useTheme} from './hooks/useTheme'
+import {useBase} from './hooks/useBase'
```
Comment (hoangSTS) [inline src/Hello.tsx:10]: Create index files to export their components

**Review by hoangSTS — CHANGES_REQUESTED** [review]

**Review by honghoangsts — CHANGES_REQUESTED** [review]

**Comment by TranQuangPhi** [comment]
please fix conflict

File: src/Hello.tsx [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+const Hello = () => {
+  return (
+    <BaseProvider theme={theme}>
+      <Welcome />
+    </BaseProvider>
+  )
+}
+
+const Welcome = () => {
```
Comment (ThinhKimVo) [inline src/Hello.tsx:30]: for short, should you move this component into one component?.

**Review by huydosgtech — APPROVED** [review]

**Pull request merged**